File: advent-of-code/2015/01/01/solution.py
from functools import reduce
from pathlib import Path


def calculate_floor(source: str) -> int:
    return reduce(lambda floor, character: 
        floor + 1 if character == '(' else 
        floor - 1 if character == ')' else 
        floor, source, 0)

# Test cases with assertions
assert calculate_floor("(())") == 0
assert calculate_floor("()()") == 0
assert calculate_floor("(((") == 3
assert calculate_floor("(()(()(") == 3
assert calculate_floor("))(((((") == 3
assert calculate_floor("())") == -1
assert calculate_floor("))(") == -1
assert calculate_floor(")))") == -3
assert calculate_floor(")())())") == -3

# Calculate and print the answer
input_file = Path(__file__).parent.parent / "input.txt"
print(f"Python: {calculate_floor(input_file.read_text())}")

File: advent-of-code/2015/01/01/solution.clj
(defn calculate-floor [source]
  (reduce (fn [floor character]
            (cond
              (= character \() (inc floor)
              (= character \)) (dec floor)
              :else floor))
          0
          source))

;; Test cases with assertions
(assert (= 0 (calculate-floor "(())")))
(assert (= 0 (calculate-floor "()()")))
(assert (= 3 (calculate-floor "(((")))
(assert (= 3 (calculate-floor "(()(()(")))
(assert (= 3 (calculate-floor "))(((((")))
(assert (= -1 (calculate-floor "())")))
(assert (= -1 (calculate-floor "))(")))
(assert (= -3 (calculate-floor ")))")))
(assert (= -3 (calculate-floor ")())())")))


;; Calculate and print the answer
(let [script-dir (.getParent (java.io.File. *file*))
      input-path (str script-dir "/../input.txt")]
  (println "Clojure:" (calculate-floor (slurp input-path))))
